lib/stemcell/launcher.rb

```diff
@@ -291,5 +297,44 @@ def try_file(opt="")
         File.read(File.expand_path(opt)) rescue opt
     end
 
+    MAX_ATTEMPT = 3
+    def initial_retry_sec
+      return 1
+    end
+
+    # Return a Hash of instance => error. Empty hash indicates "no error"
+    # for code block:
+    #   - if block returns nil, success
+    #   - if block returns non-nil value (e.g., exception), retry 3 times w/ backoff
+    #   - if block raises exception, fail
+    def run_batch_operation(instances)
+      # keep track of instances for retry
+      retryable_instances = instances.map { |i| [i, nil] }
+      errors = {}
+      attempt = 0
+
+      while !retryable_instances.empty? && attempt < MAX_ATTEMPT
+        # sleep idempotently except for the first attempt
+        sleep(initial_retry_sec * 2 ** attempt) if attempt != 0
+
+        retryable_instances.keep_if do |instance, _|
+          begin
+            yield(instance)
+          rescue => e
+            errors[instance] = e
+          end
+        end
+
+        attempt += 1
+      end
+
+      retryable_instances.each { |i, e| error[i] = e if !error.nil? }
+      errors
+    end
```

Comment:
If I'm reading this correctly, we go through all the instances and make a best-effort first pass, then if there were any errors, we circle back and take care of those, and we keep doing this until we either have all successes or run out of attempts. 

I think it would be ok to retry operations on an instance if they error out before moving onto the next instance. This would make the code simpler because we don't have to keep track of the retryable instances. Also, instead of returning a hash, I think a positional array of results is probably ok. 

This would be my suggestion for `run_batch_operation`:

```ruby
# if block returns nil, success
# if block returns non-nil value (e.g., exception), retry 3 times w/ backoff
# if block raises exception, fail
def run_batch_operation(instances)
  instances.map do |instance|
    begin
      attempt = 0
      result = nil
      while attempt < MAX_ATTEMPTS
        result = yield(instance)
        break if result.nil? # nil indicates success
        sleep(2 ** attempt)  # exponential backoff: 1s, 2s, 4s
        attempt += 1
      end
      result # result for this instance is nil or returned exception
    rescue => e
      e # result for this instance is caught exception
    end
  end
end
```